### ISSUE TITLE
Region object is not thread safe because it has NSCalendar 

### ISSUE THREAD

**Issue by ypresto** [malcommac/SwiftDate#189]
NSCalendar is not thread safe according to apple's document:
https://developer.apple.com/library/mac/documentation/Cocoa/Conceptual/Multithreading/ThreadSafetySummary/ThreadSafetySummary.html

So Region and DateInRegion is also not thread safe.
Is it expected..? If so I'd be better to be documented :)

**Comment by ypresto** [comment]
Region is exposing mutable NSCalendar instance to public, so Region is **mutable**. Is it also expected..?

```swift
    public let calendar: NSCalendar!
```

**Comment by ypresto** [comment]
Note that I want to cache default Region object because calling `nsDateInstance.add(...)` instantiates new NSCalendar object via Region and it is very slow.

**Comment by malcommac** [comment]
```calendar``` is defined as let and assigned at init time so it cannot be change further.
However there is a similar problem with ```timezone``` and ```locale``` which are var.
I think you are right, we should mark them as read-only property and the entire ```Region``` struct should be atomic.

**Comment by ypresto** [comment]
Yes, `let calendar` cannot be changed. But NSCalendar itself is mutable class. For example, locale can be changed by `region.calendar.locale = ...` from outside of the library.

**Comment by malcommac** [comment]
yeah, you are right. So what do you think is the best approach?
We can return a copy of the calendar var and keep it private. Any other solution?

**Comment by Hout** [comment]
We could create a new struct that mimics the data model of `NSCalendar`. That would create a dependency from Apple's `NSCalendar`class though. Your solution is easier.

**Comment by Hout** [comment]
On the other hand: Apple did the same with `NSArray` and `Array`: Swiftifying. Doing that with `NSCalendar` to a `Calendar` struct would imply A LOT of work though.

**Comment by Hout** [comment]
Wouldn't that be fun? :smile:

**Comment by Hout** [comment]
NSTimeZone is thread unsafe too...

Something like this?

```
protocol Calendarish {
    public let identifier: String
    public let locale: Locale
    public let timeZone: TimeZone
    ...
}

struct GregorianCalendar : Calendarish {
    ...
}

struct TimeZone {
    public let abbreviation String
    public let GMTOffset NSTimeInterval
    ...
}
```

**Comment by ypresto** [comment]
> https://developer.apple.com/library/mac/documentation/Cocoa/Conceptual/Multithreading/ThreadSafetySummary/ThreadSafetySummary.html

NSTimeZone is thread safe according to the document.

> struct that mimics the data model of NSCalendar

I agree with this, because only public constructor `Region(calendarName:, timeZoneName:, localeName:)` does not depend on NSCalendar instance.

**Comment by malcommac** [comment]
I'm a bit afraid to make another set of classes where the only scope is to avoid this issue; we will end with a big library of "non necessary things".
We could hide the calendar inside our region and expose it as a copy; what's wrong with this propose?

**Comment by Hout** [comment]
Woops I looked in the wrong list and stand corrected. ;-)

**Comment by Hout** [comment]
@malcommac nothing wrong with your approach. I am just playing with the idea.

**Comment by Hout** [comment]
Is [this](http://ddeville.me/2011/02/creating-shared-instances-of-non-thread-safe-classes/) an idea for making calendars thread safe?

**Comment by malcommac** [comment]
yeah sure :) btw I can't see benefits; at the start of this project my idea is to completely hide the cocoa framework behind and make something with new structs and classes
However a day you said to me we should keep things simple and in the majority of cases people love to use standard classes so with SwiftDate you can still use NSDate and move on DateInRegion only if you need to play with a custom configuration. It was enlightening and now I think adding tons of "fake" classes only to hide a behaviour is not the best solution. I'm afraid of that.

**Comment by malcommac** [comment]
@Hout We are using this for our NSDateFormatter object too!

**Comment by ypresto** [comment]
I meant like below:

```swift
class Region {
    let calendarName: CalendarName
    let timeZoneName: TimeZoneName
    let localName: LocaleName

    var calendar: NSCalendar {
        // It might be better to cache calendar per Region instance instead of global one.
        return CalendarCache.sharedCalendarForCurrentThtread(region: self)
    }

    ...
}
```

This does not create any new abstraction or mimics, and just make Region thread safe and cachable..!

**Comment by Hout** [comment]
@malcommac why aren't we as smart as @ypresto ? ;-)

**Comment by ypresto** [comment]
Note that `var calendar` is internal access, and cannot be manipulated from outside of instance.
Cost of copying NSCalendar is not ignorable, so I'd be better to copy it only when necessary (for use from outside of the library).

```swift
func newCalendar() -> NSCalendar {
    ...
}
```

**Comment by Hout** [comment]
@malcommac I think you are correct about the origins of your project and the decision direction.

I got carried away; give me a hint and I have a thousand ideas ;-) This Calendar idea originates from the expectation that over time the Cocoa classes will be Swiftified in the swift.org project. A new and open sourced Calendar struct would be an enlightment. It would be a project on its ownthough  and way passed the principles of this project.

So how do we progress here?

**Comment by Hout** [comment]
I will take on this one with @ypresto 's idea for defining an immutable Region struct.

**Comment by Sega-Zero** [comment]
There is an accepted [proposal](https://github.com/apple/swift-evolution/blob/master/proposals/0069-swift-mutability-for-foundation.md) for swift 3 that will make immutable `Calendar` wrapper over `NSCalendar`. It is better to wait for it, I think.

**Comment by Hout** [comment]
Great find Sergey! I agree to wait for Swift 3. Do you concur @malcolmmac?
